### PR TITLE
chore(private_dns): tag dns zone and delete only if owned

### LINKED ISSUE #1749: BYO DNS Zone
/kind feature

**What steps did you take and what happened:**
1. Created a private cluster with custom vnet
2. Create a dns zone through azure cli/ui and the appropriate links and records 
3. Delete the cluster
4. capz tries to delete the dns zone created in step 2, but fails because there are some active links.

**What did you expect to happen:**
CAPZ should support byo dns zone

**Anything else you would like to add:**

**Environment:**

- cluster-api-provider-azure version: 
- Kubernetes version: (use `kubectl version`): 
- OS (e.g. from `/etc/os-release`):

### PR DESCRIPTION
**Pull request by sonasingh46**
**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:
This PR tags private DNS zone created by CAPZ and delete
only if it has been created by CAPZ.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #1749 

**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**TODOs**:

- [ ] squashed commits
- [ ] includes documentation
- [ ] adds unit tests

**Release note**:

```release-note
None
```

Signed-off-by: Ashutosh Kumar <sonasingh46@gmail.com>

### CODE REVIEW

**Review by CecileRobertMichon — COMMENTED** [review]
/ok-to-test

File: azure/services/privatedns/privatedns.go [hunk 1]
```diff
@@ -59,7 +61,15 @@ func (s *Service) Reconcile(ctx context.Context) error {
...
-		err := s.client.CreateOrUpdateZone(ctx, s.Scope.ResourceGroup(), zoneSpec.ZoneName, privatedns.PrivateZone{Location: to.StringPtr(azure.Global)})
+		pDNS := privatedns.PrivateZone{
+			Location: to.StringPtr(azure.Global),
+			Tags: converters.TagsToMap(infrav1.Build(infrav1.BuildParams{
+				ClusterName: s.Scope.ClusterName(),
+				Lifecycle:   infrav1.ResourceLifecycleOwned,
+				Additional:  s.Scope.AdditionalTags(),
+			})),
+		}
+		err := s.client.CreateOrUpdateZone(ctx, s.Scope.ResourceGroup(), zoneSpec.ZoneName, pDNS)
```
Comment (CecileRobertMichon) [inline azure/services/privatedns/privatedns.go:72]: Wouldn't this cause us to add the tags to any pre-existing DNS zone, including a user's BYO zone?

File: azure/services/privatedns/privatedns.go [hunk 2]
```diff
@@ -59,7 +61,15 @@ func (s *Service) Reconcile(ctx context.Context) error {
...
 		if err != nil {
 			return errors.Wrapf(err, "failed to create private DNS zone %s", zoneSpec.ZoneName)
 		}
```
Comment (CecileRobertMichon) [inline azure/services/privatedns/privatedns.go:75]: we should consider also not creating/deleting vnet links if both the DNS zone and the vnet are not managed by capz

File: azure/services/privatedns/privatedns.go (hunk 1 above)
Comment (sonasingh46) [reply azure/services/privatedns/privatedns.go:72]: This is good point. I was thinking on the similar lines. But when I looked at piublicip reconcile there was no such handling in the Reconcile for managed vs unmanaged. ( Not sure if it applies to that or not ) 
Ref: https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/main/azure/services/publicips/publicips.go#L83
But I will make the changes here so that BYO DNS does not get tagged ( and hence reconciled) by CAPZ

File: azure/services/privatedns/privatedns.go (hunk 2 above)
Comment (sonasingh46) [reply azure/services/privatedns/privatedns.go:89]: I was thinking, simply we should not be reconciling/deleting vnet links if the vnet is not managed. Does the management of vnet depends on wether DNS is managed or not?

Comment (sonasingh46) [reply azure/services/privatedns/privatedns.go:89]: By the way, I have raised the commit according to above comment of mine. Let me know if this is incorrect or needs changes.

File: azure/services/privatedns/privatedns.go [hunk 3]
```diff
@@ -142,3 +204,25 @@ func (s *Service) Delete(ctx context.Context) error {
 	}
 	return nil
 }
+
+// isPrivateDNSManaged returns true if the private DNS has an owned tag with the cluster name as value,
+// meaning that the DNS lifecycle is managed.
+func (s *Service) isPrivateDNSManaged(ctx context.Context, resourceGroup, zoneName string) (bool, error) {
+	gotZone, err := s.client.GetZone(ctx, resourceGroup, zoneName)
```
Comment (shysank) [inline azure/services/privatedns/privatedns.go:211]: nit
```suggestion
	zone, err := s.client.GetZone(ctx, resourceGroup, zoneName)
```

File: azure/services/privatedns/privatedns.go [hunk 4]
```diff
@@ -142,3 +204,25 @@ func (s *Service) Delete(ctx context.Context) error {
...
+		return false, err
+	}
+	tags := converters.MapToTags(gotZone.Tags)
+	return tags.HasOwned(s.Scope.ClusterName()), nil
+}
+
+// isVnetLinkManaged returns true if the vnet link has an owned tag with the cluster name as value,
+// meaning that the vnet link lifecycle is managed.
+func (s *Service) isVnetLinkManaged(ctx context.Context, resourceGroupName, zoneName, vnetLinkName string) (bool, error) {
+	gotZone, err := s.client.GetLink(ctx, resourceGroupName, zoneName, vnetLinkName)
```
Comment (shysank) [inline azure/services/privatedns/privatedns.go:222]:
```suggestion
	vnetLink, err := s.client.GetLink(ctx, resourceGroupName, zoneName, vnetLinkName)
```

File: azure/services/privatedns/client.go [hunk 5]
```diff
@@ -108,6 +120,17 @@ func (ac *azureClient) DeleteZone(ctx context.Context, resourceGroupName, name s
 	return err
 }
 
+// GetLink returns a vnet link.
+func (ac *azureClient) GetLink(ctx context.Context, resourceGroupName, zoneName, vnetLinkName string) (privatedns.VirtualNetworkLink, error) {
+	ctx, _, done := tele.StartSpanWithLogger(ctx, "privatedns.AzureClient.GetZone")
```
Comment (shysank) [inline azure/services/privatedns/client.go:125]:
```suggestion
	ctx, _, done := tele.StartSpanWithLogger(ctx, "privatedns.AzureClient.GetLink")
```

File: azure/services/privatedns/privatedns.go [hunk 6]
```diff
@@ -57,15 +59,51 @@ func (s *Service) Reconcile(ctx context.Context) error {
...
+			isManaged = true
+		}
+		if !isManaged {
+			s.Scope.V(2).Info("Skipping reconciliation of unmanaged private DNS zone", "private DNS", zoneSpec.ZoneName)
+			return nil
+		}
 		// Create the private DNS zone.
 		s.Scope.V(2).Info("creating private DNS zone", "private dns zone", zoneSpec.ZoneName)
-		err := s.client.CreateOrUpdateZone(ctx, s.Scope.ResourceGroup(), zoneSpec.ZoneName, privatedns.PrivateZone{Location: to.StringPtr(azure.Global)})
+		pDNS := privatedns.PrivateZone{
```
Comment (shysank) [inline azure/services/privatedns/privatedns.go:78]: Maybe not part of this pr, but should we call `CreateOrUpdateZone` if there is already a zone existing since we don't allow changes to any of its properties?

File: azure/services/privatedns/privatedns.go (hunk 1 above)
Comment (CecileRobertMichon) [reply azure/services/privatedns/privatedns.go]: Interesting, we should verify but I think that might be a bug. Also coincidentally would be fixed by https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/1745/files#diff-101a2cb7fa509c5208b45ff64bc92af67c470f1a3de730a5a05f60b95b7b1681R58.

Comment (CecileRobertMichon) [reply azure/services/privatedns/privatedns.go]: BTW the only issue with not tagging existing DNS zones is that we won't tag the ones created by earlier versions of CAPZ and thus it would cause the old DNS zones to be treated as BYO and not be deleted with this new code.

Comment (CecileRobertMichon) [reply azure/services/privatedns/privatedns.go]: added a comment in https://github.com/kubernetes-sigs/cluster-api-provider-azure/issues/1692

File: azure/services/privatedns/privatedns.go (hunk 6 above)
Comment (sonasingh46) [reply azure/services/privatedns/privatedns.go:78]: Opened up an issue to keep track of this one. 
https://github.com/kubernetes-sigs/cluster-api-provider-azure/issues/1872

File: azure/services/privatedns/privatedns.go (hunk 1 above)
Comment (sonasingh46) [reply azure/services/privatedns/privatedns.go]: >BTW the only issue with not tagging existing DNS zones is that we won't tag the ones created by earlier versions of CAPZ and thus it would cause the old DNS zones to be treated as BYO and not be deleted with this new code.

Shall we cover this as part of upgrade? One of the things that I can think top off my head is, before upgrading CAPZ users can annotate DNS to adopt it for reconcile. May be we can automate some of this process but there has to be an user intervention unless the DNS that is NOT a BYO has some footprint that differentiates to that of DNS that is a BYO.

Comment (sonasingh46) [reply azure/services/privatedns/privatedns.go]: Infact, IMO we can make a generic annotation key to solve similar issue for other resources or in general if we at some point of time want to adopt resources that are not reconciled or BYO 

What do you think?

Comment (shysank) [reply azure/services/privatedns/privatedns.go]: >BTW the only issue with not tagging existing DNS zones is that we won't tag the ones created by earlier versions of CAPZ and thus it would cause the old DNS zones to be treated as BYO and not be deleted with this new code.

This would only affect a subset of clusters where resource group is also not managed. Perhaps, we can add some logging for those scenarios to nudge the user in the right direction?

Comment (sonasingh46) [reply azure/services/privatedns/privatedns.go]: Sure. I am already logging for not managed DNS.

Comment (CecileRobertMichon) [reply azure/services/privatedns/privatedns.go]: let's change that log to v(1) so it's printed even if the user doesn't change verbosity and add a little blurb about adding the tag to nudge the user to add the tag if they didn't create the zone themselves? Maybe we can even point them to a link in docs if we document this feature

**Comment by CecileRobertMichon** [comment]
https://capz.sigs.k8s.io/topics/custom-dns.html would be a good place to document the feature

**Comment by CecileRobertMichon** [comment]
@sonasingh46 please squash commits

**Comment by CecileRobertMichon** [comment]
@sonasingh46 can you please rebase when you get a chance?

**Comment by sonasingh46** [comment]
@CecileRobertMichon -- Done rebase.

**Review by CecileRobertMichon — COMMENTED** [review]
/lgtm
/approve
